### PR TITLE
Feat-p22_43 editing_tennis_player_profile

### PR DESCRIPTION
**Pull request by rapaicd**

### CODE REVIEW

File: Tennis-scheduler/src/main/resources/application.properties [hunk 1]
```diff
@@ -6,7 +6,7 @@ spring.datasource.url=jdbc:postgresql://localhost:5432/TennisScheduler
 
 #Credentionals for communication with database
 spring.datasource.username=postgres
-spring.datasource.password=postgres
+spring.datasource.password=danilo
```
Comment (Dzela) [inline Tennis-scheduler/src/main/resources/application.properties]: Revert this change before merge

File: Tennis-scheduler/src/main/java/com/tennis/tennisscheduler/controllers/AuthenticationController.java [hunk 2]
```diff
@@ -35,4 +38,27 @@ public ResponseEntity<UserTokenStateDto> login(@RequestBody AuthenticationReques
 
         return ResponseEntity.ok(new UserTokenStateDto(jwt, user.getRole().getRoleName()));
     }
+
+    @PreAuthorize("hasRole('TENNIS_PLAYER')")
+    @PutMapping("/")
```
Comment (markoacimovic) [inline Tennis-scheduler/src/main/java/com/tennis/tennisscheduler/controllers/AuthenticationController.java]: Better path name please

File: Tennis-scheduler/src/main/java/com/tennis/tennisscheduler/controllers/PersonController.java [hunk 3]
```diff
@@ -65,4 +73,15 @@ public ResponseEntity<PersonDto> updatePerson(@PathVariable long id,@RequestBody
         Person person = personService.updatePerson(id, personDtoMapper.fromPersonDtoToPerson(personDto));
         return new ResponseEntity<>(personDtoMapper.fromPersonToPersonDto(person),HttpStatus.OK);
     }
+
+    @PutMapping("/")
+    @PreAuthorize("hasRole('TENNIS_PLAYER')")
+    public ResponseEntity<PersonDto> updatePerson(@RequestBody PersonDto personDto){
```
Comment (Dzela) [inline Tennis-scheduler/src/main/java/com/tennis/tennisscheduler/controllers/PersonController.java]: Why new endpoint?

File: Tennis-scheduler/src/main/java/com/tennis/tennisscheduler/controllers/AuthenticationController.java [hunk 4]
```diff
@@ -35,4 +38,27 @@ public ResponseEntity<UserTokenStateDto> login(@RequestBody AuthenticationReques
 
         return ResponseEntity.ok(new UserTokenStateDto(jwt, user.getRole().getRoleName()));
     }
+
+    @PreAuthorize("hasRole('TENNIS_PLAYER')")
```
Comment (Dzela) [inline Tennis-scheduler/src/main/java/com/tennis/tennisscheduler/controllers/AuthenticationController.java]: Why ADMIN role can't change their password?

File: Tennis-scheduler/src/main/java/com/tennis/tennisscheduler/controllers/PersonController.java [hunk 5]
```diff
@@ -1,6 +1,7 @@
 package com.tennis.tennisscheduler.controllers;
 
 import com.tennis.tennisscheduler.dtos.PersonDto;
+import com.tennis.tennisscheduler.dtos.UpdatePasswordDto;
```
Comment (Dzela) [inline Tennis-scheduler/src/main/java/com/tennis/tennisscheduler/controllers/PersonController.java]: Remove unused imports. Please check other classes as well.

File: Tennis-scheduler/src/main/java/com/tennis/tennisscheduler/dtos/PersonDto.java [hunk 6]
```diff
@@ -17,4 +17,5 @@ public class PersonDto {
     public Date birthday;
     public Address address;
     public String password;
+    public String confirmPassword;
```
Comment (Dzela) [inline Tennis-scheduler/src/main/java/com/tennis/tennisscheduler/dtos/PersonDto.java]: Where is this field used?

Comment (rapaicd) [reply Tennis-scheduler/src/main/java/com/tennis/tennisscheduler/dtos/PersonDto.java]: This field is added because of form on FE, but we didn't provide that check on BE because we have it on FE.

File: Tennis-scheduler/src/main/java/com/tennis/tennisscheduler/controllers/PersonController.java (hunk 3 above)
Comment (rapaicd) [reply Tennis-scheduler/src/main/java/com/tennis/tennisscheduler/controllers/PersonController.java]: Second one is for logged user to edit his personal information and the first one is for admin so he can edit information about any user. 
Maybe we don't need the first one endpoint?

Comment (Dzela) [reply Tennis-scheduler/src/main/java/com/tennis/tennisscheduler/controllers/PersonController.java]: The first one should be enough, both roles can use it, you can just add check for tennis_player role before calling update method to see if the user they want to update is their own or not. If yes let them do it, if not return 401 Unauthorized.

File: Tennis-scheduler/src/main/java/com/tennis/tennisscheduler/dtos/PersonDto.java (hunk 6 above)
Comment (Dzela) [reply Tennis-scheduler/src/main/java/com/tennis/tennisscheduler/dtos/PersonDto.java]: Ok, if it's not needed here please delete it.

File: Tennis-scheduler/src/main/java/com/tennis/tennisscheduler/controllers/PersonController.java [hunk 7]
```diff
@@ -59,13 +64,25 @@ public ResponseEntity<PersonDto> getPersonById(@PathVariable long id){
         return new ResponseEntity<>(personDtoMapper.fromPersonToPersonDto(person),HttpStatus.OK);
     }
     @PutMapping("/{id}")
+    @PreAuthorize("hasAnyRole('TENNIS_PLAYER','ADMIN')")
     public ResponseEntity<PersonDto> updatePerson(@PathVariable long id,@RequestBody PersonDto personDto){
+
         Person personExisting = personService.findById(id);
-        if (personExisting == null) {
```
Comment (Dzela) [inline Tennis-scheduler/src/main/java/com/tennis/tennisscheduler/controllers/PersonController.java:64]: You can leave this check here. No matter the role, if person doesn't exist in the database nothing else should be done. Also return NOT_FOUND since that is what happened, BAD_REQUEST means that request is malformed.

File: Tennis-scheduler/src/main/java/com/tennis/tennisscheduler/controllers/PersonController.java [hunk 8]
```diff
@@ -59,13 +64,25 @@ public ResponseEntity<PersonDto> getPersonById(@PathVariable long id){
...
-            return new ResponseEntity<>( HttpStatus.BAD_REQUEST);
-        }
+        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
+        Person user = (Person) authentication.getPrincipal();
 
-        Person person = personService.updatePerson(id, personDtoMapper.fromPersonDtoToPerson(personDto));
-        return new ResponseEntity<>(personDtoMapper.fromPersonToPersonDto(person),HttpStatus.OK);
+        if(id==user.getId()) {
```
Comment (Dzela) [inline Tennis-scheduler/src/main/java/com/tennis/tennisscheduler/controllers/PersonController.java]: The code in if and else if is basically the same. The only thing different is that in the first one you are using user.getId() which should be the same as id sent in the request if the TENNIS_PLAYER is trying to update their own profile. No need for duplication.
What you can do is to create separate if for checking the role and the id, like you had above for checking if person exists in the database (which you should leave as was). If the id is different return 401, if not just let it do the same thing as ADMIN.

Comment (rapaicd) [reply Tennis-scheduler/src/main/java/com/tennis/tennisscheduler/controllers/PersonController.java]: thanks for the suggestion :)

File: Tennis-scheduler/src/main/java/com/tennis/tennisscheduler/controllers/AuthenticationController.java [hunk 9]
```diff
@@ -38,6 +43,29 @@ public ResponseEntity<UserTokenStateDto> login(@RequestBody AuthenticationReques
...
+                    user.getEmail(), updatePasswordDto.newPassword));
+
+            SecurityContextHolder.getContext().setAuthentication(authentication);
+            jwt = tokenUtils.generateToken(user.getEmail());
+            message = "Successfully changed password.";
+        } else
+            message = "Current password is not correct.";
+
+
+        return new ResponseEntity<>(new UserWithChangedPasswordDto(new UserTokenStateDto(jwt, user.getRole().getRoleName()), message)
```
Comment (Dzela) [inline Tennis-scheduler/src/main/java/com/tennis/tennisscheduler/controllers/AuthenticationController.java:69]: Should not return 200 OK if password was not changed for any reason.

**Review by markoacimovic — APPROVED** [review]
